Hw11 cache engine

daria-grebneva opened the pull request.

daria-grebneva committed:
Скелет cache engine

daria-grebneva committed:
Добавила запись в кэш данных с базы

petrelevich commented on hw11-cache-engine/homework/src/main/java/ru/otus/cachehw/MyCache.java:
```diff
@@ -0,0 +1,47 @@
+public class MyCache<K, V> implements HwCache<K, V> {
+//Надо реализовать эти методы
+    private Map<K, V> cache = new WeakHashMap<>();
+    private List<HwListener> listeners = new ArrayList<>();
```
оба поля стоит сделать final

daria-grebneva replied:
поправлено

petrelevich commented on hw11-cache-engine/homework/src/main/java/ru/otus/cachehw/MyCache.java:
```diff
@@ -0,0 +1,47 @@
+
+    private void notifyListeners(K key, V value, String action) {
+        for (HwListener listener : listeners) {
+            listener.notify(key, value, action);
```
стоит добавить try-catch

daria-grebneva replied:
поправлено

petrelevich commented on hw11-cache-engine/homework/src/main/java/ru/otus/HomeWork.java:
```diff
@@ -0,0 +1,131 @@
+        var dbExecutor = new DbExecutorImpl();
+
+// Инициализация кеша
+        HwCache<Long, Client> clientCache = new MyCache<>();
```
на вебинаре показывал, почему плохо использовать Long, Integer в качестве ключа.
лучше String

daria-grebneva replied:
Добавила метод getKeyForCache() и в нем вызов intern(). Правильно ли я понимаю, что теперь при таком подходе в первый вызов getKeyForCache() не возьмется из пула строк (если мы заранее до этого данную строку не инициализировали), но если мы обратимся второй раз с тем же ключом, то значение уже возьмется из пула?
Или использование intern() в данном случае избыточно?

petrelevich replied:
intern() тут не обязательно.

daria-grebneva replied:
Поправила

petrelevich commented on hw11-cache-engine/homework/src/main/java/ru/otus/HomeWork.java:
```diff
@@ -0,0 +1,131 @@
+        managerCache.addListener(managerListener);
+
+// Работа с клиентом
+        EntityClassMetaData entityClassMetaDataClient = new EntityClassMetaDataImpl(Client.class);
```
не хватает generic-а

daria-grebneva replied:
Поправила

petrelevich commented on hw11-cache-engine/homework/src/main/java/ru/otus/HomeWork.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    private static Client getClient(HwCache<Long, Client> clientCache, DBServiceClient dbServiceClient, Long key) {
+        Client client = clientCache.get(key);
```
кеш должен быть встроен внутрь сервиса dbServiceClient.
пользователь dbServiceClient не должен самостоятельно работать с кешом, это должно быть "прозрачно".

daria-grebneva replied:
Поправила

petrelevich commented on hw11-cache-engine/homework/src/main/java/ru/otus/mapper/DataTemplateJdbc.java:
```diff
@@ -0,0 +1,131 @@
+            try {
+                while (rs.next()) {
+                    List<String> l = getFieldsValuesFromFromDb(rs);
+                    // TODO:: спросить, как данное место можно улучшить, получаю IllegalArgumentException
```
сюда
`public T newInstance(Object ... initargs)`
можно передать массив
`entityClassMetaData.getConstructor().newInstance(new Object[]{});`

daria-grebneva replied:
Спасибо! Поняла свою ошибку, Id-шник объекта должен быть частью массива

daria-grebneva committed:
Правки комментариев:
- Кеш теперь встроен внутрь dbServiceClient и dbServiceManager
- Использование String вместо Long в качества ключа для кеша

petrelevich commented on hw11-cache-engine/demo/src/main/java/ru/otus/cachehw/MyCache.java:
```diff
@@ -0,0 +1,52 @@
+public class MyCache<K, V> implements HwCache<K, V> {
+//Надо реализовать эти методы
+    private final Map<K, V> cache = new WeakHashMap<>();
+    private final List<HwListener> listeners = new ArrayList<>();
```
у HwListener не хватает дженерика.

daria-grebneva replied:
Поправлено

petrelevich commented on hw11-cache-engine/demo/src/main/java/ru/otus/cachehw/MyCache.java:
```diff
@@ -0,0 +1,52 @@
+                listener.notify(key, value, action);
+            }
+        } catch (Exception e) {
+            throw new RuntimeException("Unexpected error during notification: " + e);
```
исключения надо логировать, но не пробрасывать, чтобы не прерывать обход лисенеров.

daria-grebneva replied:
Готово)

petrelevich commented on hw11-cache-engine/demo/src/main/java/ru/otus/crm/service/DbServiceClientImpl.java:
```diff
@@ -0,0 +1,97 @@
+    @Override
+    public Client saveClient(Client client) {
+        try {
+            clientCache.addListener(clientListener);
```
добавлять лисенер тут не надо.

daria-grebneva replied:
Поправлено

petrelevich commented on hw11-cache-engine/demo/src/main/java/ru/otus/crm/service/DbServiceClientImpl.java:
```diff
@@ -0,0 +1,97 @@
+                }
+                dataTemplate.update(connection, client);
+                log.info("updated client: {}", client);
+                return client;
```
созданного и, особенно важно, измененного клиента надо обязательно положить в кеш.

daria-grebneva replied:
Поправила

daria-grebneva committed:
Правки комментариев

daria-grebneva committed:
Убрала эксепшен при логировании, чтобы не прерывать обход листенеров

daria-grebneva committed:
Merge branch 'main' into hw11-cache-engine
